src/tabs/presets/PresetsRepoIndexed/PresetParser.js

```diff
@@ -0,0 +1,224 @@
+'use strict';
+
+class PresetParser {
```

Comment:
A bit of nitpick, but why does every method modify the passed in arguments? This will lead to maintenance hell for other people.

Also unit testing would be much easier if it's just data in data out.

Reply:
There are only 4 public methods in this "class":
`constructor(settings)`
`readPresetProperties(preset, strings)`
`removeUncheckedOptions(strings, checkedOptions)`
`isIncludeFound(strings)`

only `readPresetProperties(preset, strings)` actually modifies the argument - `preset`.
"Preset" is a big structure that describes a preset with all its properties/attributes and even CLI text at some point.
This function parses the preset text and sets preset properties according to this text.
Properties like `preset.description`,` preset.warning`, `preset.discussion` etc.

Most of the other "private" functions in this class follow the same "pattern" - they only modify/add properties within `preset`, all other arguments remain untouched.

Reply:
But all of those are connected and used only once, in the end reading this code we still need to jump to the methods and see what they actually do to the arguments as they are used once and then just chained. Then there is a mix of mutating and none mutating methods.

Just feels counter intuitive as it needs either knowledge of what it does inside or going and reading every step which modifies the data 🤷‍♂️

Reply:
Yeah, that's not my easiest file to read. But it actually does a heavy job - parsing our own little "programming" language of preset attributes, options, options groups etc.

Reply:
Did we define this language anywhere? Might be an interesting task to write proper AST parser instead 😉

Reply:
Yes Sir. The most detailed description is here, made by Ctzsnooze
https://github.com/betaflight/firmware-presets

it is pretty detailed, only missing a couple of most recent  things:
`#$ PRIORITY: ....` and `#$ OPTION_GROUP BEGIN/END`

Reply:
Now the most `full` parser and syntax checker are here:
https://github.com/betaflight/firmware-presets/tree/master/indexer
Here in the Configurator its a simplified version without syntax check and for only necessary attributes.